packages/itwinui-react/src/core/Panels/Panels.tsx

```diff
@@ -0,0 +1,488 @@
+
+// #region PanelsWrapper
+
+export type PanelsWrapperProps = {
+  /**
+   * The initialPanel that is displayed.
+   */
+  initialActiveId: string;
+  onActiveIdChange?: (newActiveId: string) => void;
+  children: React.ReactNode;
+  /**
+   * Pass an instance created by `useInstance` to control the panels imperatively.
+   *
+   * @example
+   * const panels = Panels.useInstance();
+   * <Panels instance={panels} />
+   */
+  instance?: PanelsInstance;
+};
+
+/**
+ * Component that manages the logic for layered panels.
+ * It can be used anywhere to create layers. E.g. `Menu`, `InformationPanel`, `Popover`, etc.
+ *
+ * Notes:
+ * - A panel can have only one trigger pointing to it. i.e. out of all the triggers across all panels,
+ *   only one can point to a particular panel.
+ *
+ * @example
+ * <Panels.Wrapper
+ *   initialActiveId={panelIdRoot}
+ *   as={Surface}
+ *   style={{ inlineSize: '300px', blockSize: '500px' }}
+ * >
+ *   <Panels.Panel
+ *     id={panelIdRoot}
+ *     as={Surface}
+ *     border={false}
+ *     elevation={0}
+ *   >
+ *     <Surface.Header as={Panels.Header}>Root</Surface.Header>
+ *     <Surface.Body as={List}>
+ *       <ListItem>
+ *         <Panels.Trigger for={panelIdMoreInfo}>
+ *           <ListItem.Action>More details</ListItem.Action>
+ *         </Panels.Trigger>
+ *       </ListItem>
+ *     </Surface.Body>
+ *   </Panels.Panel>
+ *
+ *   <Panels.Panel
+ *     id={panelIdMoreInfo}
+ *     as={Surface}
+ *     border={false}
+ *     elevation={0}
+ *   >
+ *     <Surface.Header as={Panels.Header}>More details</Surface.Header>
+ *     <Surface.Body isPadded>
+ *       <Text>Content</Text>
+ *     </Surface.Body>
+ *   </Panels.Panel>
+ * </Panels.Wrapper>
+ */
+const PanelsWrapper = React.forwardRef((props, forwardedRef) => {
+  const {
+    initialActiveId,
+    children,
+    className,
+    onActiveIdChange,
+    instance: instanceProp,
+    ...rest
+  } = props;
+
+  const instanceBackup = Panels.useInstance();
+  const instance = instanceProp || instanceBackup;
+
+  const ref = React.useRef<HTMLDivElement | null>(null);
+
+  const [activePanel, _setActivePanel] = React.useState(initialActiveId);
+  const setActivePanel = React.useCallback(
+    (newActivePanel: typeof activePanel) => {
+      _setActivePanel(newActivePanel);
+      onActiveIdChange?.(newActivePanel);
+    },
+    [onActiveIdChange],
+  );
+
+  const previousActivePanel = useDelayed(activePanel);
+
+  const [_panelElements, setPanelElements] = React.useState<
+    Record<string, HTMLElement | null>
+  >({});
+  const panelElements = useLatestRef(_panelElements);
+
+  const [triggers, setTriggers] = React.useState<
+    Record<string, TriggerMapEntry>
+  >({});
+  const triggersRef = useLatestRef(triggers);
+
+  const changeActivePanel = React.useCallback(
+    async (newActiveId: string) => {
+      if (
+        // Panel transition in progress
+        activePanel !== previousActivePanel ||
```

Comment:
Continued from https://github.com/iTwin/iTwinUI/pull/2001#discussion_r1744044480: this check is not necessary, and the `previousActivePanel` state does not need to be stored in context because it is not "real" state - it's derived from `activePanel` in the one place where it is actually needed.

Reply:
> code duplication is alright here?

I don't see any code duplication? If you're talking about the fact that hook will be called separately for every individual `Panel`, then yes I think that's ok because it makes sense this state is only used by the panel and this code is not expensive to re-run.